graphql_compiler/compiler/ir_lowering_match.py

```diff
@@ -578,6 +593,194 @@ def prune_output_blocks_in_compound_match_query(compound_match_query):
         return CompoundMatchQuery(match_queries=match_queries)
 
 
+def _construct_location_to_predicate_list(match_query):
+    """Return a dict mapping location -> list of filter predicates applied at that location."""
+    location_to_predicates = {}
+    # Construct a dictionary mapping locations --> a list of predicates
+    # applied to the corresponding location (in `where_blocks`)
+    for match_traversal in match_query.match_traversals:
+        for match_step in match_traversal:
+            current_filter = match_step.where_block
+            if current_filter:
+                current_location = match_step.as_block.location
+                location_to_predicates.setdefault(current_location, []).append(
+                    current_filter.predicate)
+    return location_to_predicates
+
+
+def _predicate_list_to_where_block(predicate_list):
+    """Convert a list of predicates to an Expression that is the conjunction of all of them."""
+    if not isinstance(predicate_list, list):
+        raise AssertionError(u'Expected `list`, Received {}.'.format(predicate_list))
+    if not predicate_list:
+        return None
+
+    if not isinstance(predicate_list[0], Expression):
+        raise AssertionError(u'Non-predicate object {} found in predicate_list'
+                             .format(predicate_list[0]))
+    if len(predicate_list) == 1:
+        return predicate_list[0]
+    else:
+        return BinaryComposition(u'&&',
+                                 _predicate_list_to_where_block(predicate_list[1:]),
+                                 predicate_list[0])
+
+
+def _collect_filters_to_first_location_in_match_traversal(match_traversal, location_to_predicates):
+    """Compose all filters for a specific location into its first occurence in given traversal."""
```

Comment:
This docstring is confusing to me, I don't understand what's going on. Can you write a full docstring, with args/returns sections and everything?